framework/Optimizers/SPSA.py

```diff
@@ -63,17 +63,31 @@ def localInputAndChecks(self, xmlNode):
     self.paramDict['alpha'] = float(self.paramDict.get('alpha', 0.602))
     self.paramDict['gamma'] = float(self.paramDict.get('gamma', 0.101))
     self.paramDict['A']     = float(self.paramDict.get('A', self.limit['mdlEval']/10))
-    self.paramDict['a']     = float(self.paramDict.get('a', 0.16))
+    self.paramDict['a']     = self.paramDict.get('a', None)
     self.paramDict['c']     = float(self.paramDict.get('c', 0.005))
+    #FIXME the optimization parameters should probably all operate ONLY on normalized data!
+    #  -> perhaps the whole optimizer should only work on optimized data.
+
+    #FIXME normalizing doesn't seem to have the desired effect, currently; it makes the step size very small (for large scales)
+    #if "a" was defaulted, use the average scale of the input space.
+    #This is the suggested value from the paper, missing a 1/gradient term since we don't know it yet.
+    if self.paramDict['a'] is None:
+      #minDesiredStep = min(self.optVarsInit['ranges'].values())/10.
+      #self.paramDict['a'] = minDesiredStep #*(self.paramDict['A']+1.)**self.paramDict['alpha'] <-- without G to scale, this is bad
+      self.paramDict['a'] = np.sqrt(sum(d*d for d in self.optVarsInit['ranges'].values())) #should be normalized!
+      self.raiseAMessage('Defaulting "a" gradient parameter to',self.paramDict['a'])
+    else:
+      self.paramDict['a'] = float(self.paramDict['a'])
 
     # Normalize the parameters...
     if self.gradDict['normalize']:
-      tempMax = -1
-      for var in self.optVars:
-        if self.optVarsInit['upperBound'][var]-self.optVarsInit['lowerBound'][var] > tempMax:
-          tempMax = self.optVarsInit['upperBound'][var]-self.optVarsInit['lowerBound'][var]
-      self.paramDict['c'] = copy.deepcopy(self.paramDict['c']/tempMax)
-      self.paramDict['a'] = copy.deepcopy(self.paramDict['a']/(tempMax**2))
+      maxVarRange = max(self.optVarsInit['ranges'][var] for var in self.optVars)
+      #tempMax = -1
+      #for var in self.optVars:
+      #  if self.optVarsInit['ranges'][var] > tempMax:
+      #    tempMax = self.optVarsInit['ranges'][var]
+      self.paramDict['c'] = copy.deepcopy(self.paramDict['c']/maxVarRange) #FIXME why are these deepcopied?
+      self.paramDict['a'] = copy.deepcopy(self.paramDict['a']/(maxVarRange**2)) #FIXME why are these deepcopied? And why is this normalized to the square?
```

Comment:
f633bf6c1214c66699004b371179a04a4a0600fd is the commit where they were first normalized.